### PR TITLE
Reorganize gallery examples

### LINKED ISSUE #876: Better categories for gallery examples
The [Gallery](https://www.pygmt.org/dev/gallery/index.html) is generated from Python scripts in the [`examples/gallery`](https://github.com/GenericMappingTools/pygmt/tree/master/examples/gallery) directory.

Currently, we have 19 gallery examples in 4 directories (categories).

- Lines: 2 examples
- Coastlines and boundaries: 3 examples
- Plotting map items: 12 examples
- Grids: 2 examples

IMHO, the current categories are not good, especially that the "Plotting map items" category is too general.

We need to think about how to categorize these existing and future examples. We can check the [matplotlib gallery](https://matplotlib.org/stable/gallery/index.html) for some inspirations.

### PR DESCRIPTION
**Pull request by willschlitzer**
This pull request reorganizes the gallery examples to have more specific categories.

Fixes #876

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
@willschlitzer Thanks for opening the PR. It would be much easier to discuss how we want to group gallery examples by viewing what any new designs look like.

To make it work, you also need to update the following lines in `doc/conf.py`:

https://github.com/GenericMappingTools/pygmt/blob/55c6176a4bd4d6be1baf2c955217496205fe814d/doc/conf.py#L72-L85.

**Comment by seisman** [comment]
Ping @GenericMappingTools/python @GenericMappingTools/python-maintainers for comments.

**Comment by meghanrjones** [comment]
Looks great! Based on #967 and the great response to @liamtoney's [ROSES tutorial](https://www.youtube.com/watch?v=SSIGJEe0BIk), I hope we can eventually have a whole seismology gallery section!

**Comment by seisman** [comment]
Perhaps rename the `gallery/line` to `gallery/lines`?

File: examples/gallery/line/README.txt
Comment (seisman) [inline examples/gallery/line/README.txt:1]: @willschlitzer Need to rename the directory "line" to "lines"

**Comment by weiji14** [comment]
Kinda unrelated, but can we rename one of the `inset.py` files, one which is for the gallery and one for the tutorial, because there is a warning message:

 ```
`WARNING: Duplicate example file name(s) found. Having duplicate file names will break some links. List of files: ['../examples/tutorials/inset.py']
```

**Comment by core-man** [comment]
Do [Colorbar](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/maps/colorbar.html#sphx-glr-gallery-maps-colorbar-py) and [Legend](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/maps/legend.html#sphx-glr-gallery-maps-legend-py) only belong to [Map and map elements](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/index.html#maps-and-map-elements)? But I've no better category for them. Shall we add a category called `Additional figure elements` or `Unclarified`?

There is a category called [Images, symbols, and markers](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/index.html#images-symbols-and-markers). Shall we move `Images` to an individual category? If yes, which ones are images in the current category, [Images or EPS files on maps](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/symbols/image.html#sphx-glr-gallery-symbols-image-py)? By the way, shall we rename this gallery to something like "Images or EPS files on figures" since we can put them on any figure instead of only a map, right?

**Comment by seisman** [comment]
> There is a category called [Images, symbols, and markers](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/index.html#images-symbols-and-markers). Shall we move `Images` to an individual category? If yes, which ones are images in the current category, [Images or EPS files on maps](https://pygmt-qneb8okqg-gmt.vercel.app/gallery/symbols/image.html#sphx-glr-gallery-symbols-image-py)?

I agree we should have a separate category for images. As I commented in #876, I think the category name could be "Images and contours" or "Images, contours and fields", like matplotlib (https://matplotlib.org/stable/gallery/index.html#images-contours-and-fields).

> By the way, shall we rename this gallery to something like `Images or EPS files on figures` since we can put them on any figure instead of only a map, right?

Sounds good. Actually, `Figure.image()` can recognize most image formats (png, jpg, eps, pdf et al.) via GDAL. So I think we could just say "Put images on figures" or similar.

**Comment by core-man** [comment]
Can we control the figure order in each category? For example, in `Maps and map elements` gallery, the order is now: `Political Boundaries`, `Colorbar`, `Inset`, ..., `Sampling along tracks`. It seems that we can have a better order so that some related examples can be put together.

**Comment by willschlitzer** [comment]
> Kinda unrelated, but can we rename one of the `inset.py` files, one which is for the gallery and one for the tutorial, because there is a warning message:
> 
> ```
> `WARNING: Duplicate example file name(s) found. Having duplicate file names will break some links. List of files: ['../examples/tutorials/inset.py']
> ```

Done in 965dbf3

File: examples/gallery/line/README.txt
Comment (willschlitzer) [reply examples/gallery/line/README.txt]: Done in 3639b50. Looks like PyCharm had some questions about what to change when I hit "refactor" and I never clicked through.

**Comment by seisman** [comment]
> Can we control the figure order in each category? For example, in `Maps and map elements` gallery, the order is now: `Political Boundaries`, `Colorbar`, `Inset`, ..., `Sampling along tracks`. It seems that we can have a better order so that some related examples can be put together, e.g., `Political Boundaries`, `Color land and water`, `Shorelines` seem related to each other.

There are different ways to sort the examples in one section (see https://sphinx-gallery.github.io/stable/configuration.html#within-gallery-order).
Currently, they are sorted by file names, but perhaps it makes more sense to sort by Title (`ExampleTitleSortKey`)?

https://github.com/GenericMappingTools/pygmt/blob/ffe0c2a7b742ed24576817aa1645d03417996ef9/doc/conf.py#L91

**Comment by willschlitzer** [comment]
> > Can we control the figure order in each category? For example, in `Maps and map elements` gallery, the order is now: `Political Boundaries`, `Colorbar`, `Inset`, ..., `Sampling along tracks`. It seems that we can have a better order so that some related examples can be put together, e.g., `Political Boundaries`, `Color land and water`, `Shorelines` seem related to each other.
> 
> There are different ways to sort the examples in one section (see https://sphinx-gallery.github.io/stable/configuration.html#within-gallery-order).
> Currently, they are sorted by file names, but perhaps it makes more sense to sort by Title (`ExampleTitleSortKey`)?
> 
> https://github.com/GenericMappingTools/pygmt/blob/ffe0c2a7b742ed24576817aa1645d03417996ef9/doc/conf.py#L91

Updated in c317fb7

File: doc/conf.py
Comment (seisman) [inline doc/conf.py:78]: Any reason to put "images" in the first place? The following orders make more sense to me,

```suggestion
            "../examples/gallery/maps",
            "../examples/gallery/lines",
            "../examples/gallery/symbols",
            "../examples/gallery/images",
            "../examples/gallery/3d-plots",
```

Comment (core-man) [reply doc/conf.py:78]: > Any reason to put "images" in the first place? The following orders make more sense to me,

Yes, this order is more sense to me, too.

**Comment by seisman** [comment]
> Kinda unrelated, but can we rename one of the `inset.py` files, one which is for the gallery and one for the tutorial, because there is a warning message:
> 
> ```
> `WARNING: Duplicate example file name(s) found. Having duplicate file names will break some links. List of files: ['../examples/tutorials/inset.py']
> ```

I think we should keep the "tutorials/inset.py" but rename the `inset.py` in the gallery, because we will definitely add at least one more inset example.

File: doc/conf.py
Comment (willschlitzer) [reply doc/conf.py:78]: I was organizing it alphabetically; I'll make the change.

**Comment by core-man** [comment]
> > By the way, shall we rename this gallery to something like `Images or EPS files on figures` since we can put them on any figure instead of only a map, right?
> 
> Sounds good. Actually, `Figure.image()` can recognize most image formats (png, jpg, eps, pdf et al.) via GDAL. So I think we could just say "Put images on figures" or similar.

@willschlitzer how do you think?

File: examples/gallery/symbols/README.txt
Comment (seisman) [inline examples/gallery/symbols/README.txt:2]:
```suggestion
Symbols and markers
-------------------
```

**Comment by seisman** [comment]
> > > By the way, shall we rename this gallery to something like `Images or EPS files on figures` since we can put them on any figure instead of only a map, right?
> > 
> > 
> > Sounds good. Actually, `Figure.image()` can recognize most image formats (png, jpg, eps, pdf et al.) via GDAL. So I think we could just say "Put images on figures" or similar.
> 
> @willschlitzer how do you think? I am okay with "Put images on figures"

Let's just focus on the reorganization of the examples. We can make changes after we finish this PR.

**Comment by seisman** [comment]
Some suggestions:

1. Move `symbols/logo.py` to `images` unless you think the GMT logo is a special symbol
2. Move `maps/track_sampling.py` to `images`. @weiji14 What do you think?
3. We inconsistently use hyphens `-` and underscore `_` in file names. Google recommend hyphens (https://developers.google.com/search/docs/advanced/guidelines/url-structure).

**Comment by willschlitzer** [comment]
> Some suggestions:
> 
>     1. Move `symbols/logo.py` to `images` unless you think the GMT logo is a special symbol
> 
>     2. Move `maps/track_sampling.py` to `images`. @weiji14 What do you think?

Done in b2d21a3

> 
>     3. We inconsistently use hyphens `-` and underscore `_` in file names. Google recommend hyphens (https://developers.google.com/search/docs/advanced/guidelines/url-structure).

Done in 4b1bb78

**Review by seisman — APPROVED** [review]
This gallery now looks good to me. We can add more categories after we have examples, as discussed in #876.

Ping @GenericMappingTools/python @GenericMappingTools/python-contributors @GenericMappingTools/python-maintainers for comments.

**Comment by core-man** [comment]
Great, I am okay with this PR.

**Comment by meghanrjones** [comment]
> This gallery now looks good to me. We can add more categories after we have more examples, as discussed in #876.
> 
> Ping @GenericMappingTools/python @GenericMappingTools/python-contributors @GenericMappingTools/python-maintainers for comments.

I still don't really like colorbar and legend being in 'maps and map elements' because they use Cartesian coordinate transformations and therefore are not maps or map elements. I would prefer a separate, 'Plot embellishments' category.

**Comment by willschlitzer** [comment]
> > This gallery now looks good to me. We can add more categories after we have more examples, as discussed in #876.
> > Ping @GenericMappingTools/python @GenericMappingTools/python-contributors @GenericMappingTools/python-maintainers for comments.
> 
> I still don't really like colorbar and legend being in 'maps and map elements' because they use Cartesian coordinate transformations and therefore are not maps or map elements. I would prefer a separate, 'Plot embellishments' category.

What about moving them to the `images` folder? I know they may not be the most "image-like" plotting elements, but I think they fit best there, and I want to avoid adding more small gallery sections.

**Review by weiji14 — CHANGES_REQUESTED** [review]
> We inconsistently use hyphens `-` and underscore `_` in file names. Google recommend hyphens (https://developers.google.com/search/docs/advanced/guidelines/url-structure).

That hyphen recommendation is for URLs, not Python files. PEP8 recommends underscores for *.py files (see https://www.python.org/dev/peps/pep-0008/#package-and-module-names). Also see https://stackoverflow.com/questions/2740026/why-are-underscores-better-than-hyphens-for-file-names. I'm ok with the categories, but think that we need a bit more of a discussion on the filenames before merging this, is there a way to keep the file_names with underscores and have hyphens in the generated URLs?

**Comment by willschlitzer** [comment]
I think I made my changes in line with what @seisman and @meghanrjones discussed. I didn't include all of the categories in @seisman's comment, since we don't have gallery examples for them yet (base maps, sub plots, histograms, etc.)

**Comment by seisman** [comment]
I just pushed a few changes to this branch:

- f9536322f5ea70052d3de5d26803c28bbf27b61a: Rename two python files using underscores
- e0e507dc665626d4c6b3dac6278556592f0a09ab: Rename "inset-tutorial.py" to "insets.py"
- f6622f8d4ee3934eb631262317b7e9fcadc583cc: Fix broken references due to recent renamings
- e3c8c10aca3bfe94120966f6179b39e83b689336: Rename the directory "3d-plots" to "3d_plots" (also use underscores for directories)

**Review by seisman — APPROVED** [review]
This PR now looks good to me.

Ping @GenericMappingTools/python @GenericMappingTools/python-contributors 
@GenericMappingTools/python-maintainers for comments and approvals.

**Review by meghanrjones — APPROVED** [review]
Nice!

**Review by michaelgrund — APPROVED** [review]
Looks good! No further suggestions.

**Review by core-man — APPROVED** [review]

**Review by core-man — APPROVED** [review]

**Comment by willschlitzer** [comment]
@weiji14 Can you confirm that you are no longer requesting changes?

**Comment by seisman** [comment]
Ping @weiji14 to give this PR a final review.

File: examples/gallery/images/README.txt
Comment (weiji14) [inline examples/gallery/images/README.txt:1]: image

Just on this, I don't quite get the 'contours' and 'fields' part which seem to be from matplotlib. There's no contours nor fields involved here, so could we just do 'Images and grids' or something else?

**Review by weiji14 — COMMENTED** [review]
Oops, thanks for the ping. Juggling too much on my plate recently (handling a release on another Python package and uni stuff). Just one comment.

File: examples/gallery/images/README.txt
Comment (seisman) [reply examples/gallery/images/README.txt:1]: See my comment at https://github.com/GenericMappingTools/pygmt/issues/876#issuecomment-791649980.

> This category mainly focuses on the illustration of matrix data (e.g., images and grids)
> 
> - grdimage: plot a grid by color-coded pixels
> - grdvector: plot a grid by vectors or fields
> - grdcontour: plot a grid by contours
> - image: raster image is just a special matrix data
> - contour: never used it, but I think it's similar to grdcontour

If we follow the categories of my comment, then we will have "fields" examples in this category.

Comment (weiji14) [reply examples/gallery/images/README.txt:1]: Ah ok, so just having it as a placeholder for future examples then. Sounds good.

**Review by weiji14 — APPROVED** [review]

**Pull request merged**
